pkg/envoy/ads/cache_stream.go

```diff
@@ -3,32 +3,137 @@ package ads
+		resources, err := handler(s.catalog, proxy, s.certManager, s.proxyRegistry)
+		xdsPathTimeTrack(startedAt, typeURI, proxy, err == nil)
+		if err != nil {
+			log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrGeneratingReqResource)).Str("proxy", proxy.String()).
+				Msgf("Error generating response for typeURI: %s", typeURI.Short())
+			xdsPathTimeTrack(time.Now(), envoy.TypeADS, proxy, false)
+			return nil, err
+		}
+
+		cacheResourceMap[typeURI.String()] = resources
+	}
+
+	xdsPathTimeTrack(time.Now(), envoy.TypeADS, proxy, true)
+	return cacheResourceMap, nil
+}
+
+// ServeResources stores a group of resources as a new Snapshot with a new version in the cache.
 // It also runs a consistency check on the snapshot (will warn if there are missing resources referenced in
 // the snapshot)
-func (s *Server) RecordFullSnapshot(proxy *envoy.Proxy, snapshotResources map[string][]types.Resource) error {
-	snapshot, err := cache.NewSnapshot(
-		fmt.Sprintf("%d", s.configVersion[proxy.UUID.String()]),
-		snapshotResources,
-	)
+func (s *Server) ServeResources(proxy *envoy.Proxy, snapshotResources map[string][]types.Resource) error {
+	uuid := proxy.UUID.String()
+
+	s.configVerMutex.Lock()
+	s.configVersion[uuid]++
+	configVersion := s.configVersion[uuid]
+	s.configVerMutex.Unlock()
+
+	snapshot, err := cache.NewSnapshot(fmt.Sprintf("%d", configVersion), snapshotResources)
 	if err != nil {
 		return err
 	}
 
 	if err := snapshot.Consistent(); err != nil {
-		log.Warn().Err(err).Str("proxy", proxy.String()).Msgf("Snapshot for proxy not consistent")
+		return err
 	}
 
-	s.configVerMutex.Lock()
-	defer s.configVerMutex.Unlock()
-	s.configVersion[proxy.UUID.String()]++
+	return s.snapshotCache.SetSnapshot(context.TODO(), uuid, snapshot)
+}
+
+func getCertificateCommonNameMeta(cn certificate.CommonName) (envoy.ProxyKind, uuid.UUID, identity.ServiceIdentity, error) {
+	// XDS cert CN is of the form <proxy-UUID>.<kind>.<proxy-identity>.<trust-domain>
+	chunks := strings.SplitN(cn.String(), constants.DomainDelimiter, 5)
+	if len(chunks) < 4 {
+		return "", uuid.UUID{}, "", errInvalidCertificateCN
+	}
+	proxyUUID, err := uuid.Parse(chunks[0])
+	if err != nil {
+		log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrParsingXDSCertCN)).
+			Msgf("Error parsing %s into uuid.UUID", chunks[0])
+		return "", uuid.UUID{}, "", err
+	}
+
+	switch {
+	case chunks[1] == "":
+		return "", uuid.UUID{}, "", errInvalidCertificateCN
+	case chunks[2] == "":
+		return "", uuid.UUID{}, "", errInvalidCertificateCN
+	case chunks[3] == "":
+		return "", uuid.UUID{}, "", errInvalidCertificateCN
+	}
+
+	return envoy.ProxyKind(chunks[1]), proxyUUID, identity.New(chunks[2], chunks[3]), nil
+}
+
+// recordPodMetadata records pod metadata and verifies the certificate issued for this pod
+// is for the same service account as seen on the pod's service account
+func (s *Server) recordPodMetadata(p *envoy.Proxy) error {
```

Comment:
guess this won't be needed since https://github.com/openservicemesh/osm/pull/5058 is open

Reply:
yup, rebased and removed